### ISSUE TITLE
[CP] Fixes to infrastructure exposed by work on workspaces for the Dart SDK

### ISSUE THREAD

**Issue by athomas** [dart-lang/sdk#59739]
### Commit(s) to merge

e8dafedc2ddd0cbaf216c694312735069e317ab5
b9188c2a57af7594832038d2f66b8ad97d0b3de3

### Target

beta,stable

### Prepared changelist for beta/stable

stable:
https://dart-review.googlesource.com/c/sdk/+/401460
https://dart-review.googlesource.com/c/sdk/+/401421

beta:
https://dart-review.googlesource.com/c/sdk/+/401461
https://dart-review.googlesource.com/c/sdk/+/401422

### Issue Description

A pending CL left bot caches polluted with package_config.json files with a language version on the root project in the Dart SDK. This is a problem because this package_config.json was picked up by the package_config.json generator and failed when switching between branches with different Dart SDK versions. In particular, this broke stable tryjobs. While the bots do get recycled, this reduces our confidence in CPs and therefore should be considered a release blocker.

### What is the fix

Run the generator with a hardcoded empty package config json and fix another issue connected to branch switching when git leaves empty directories in the builder cache (exposed by recent package monorepo moves).

### Why cherry-pick

Release blocker.

### Risk

Low, infra code only that is a no-op unless the build is broken (where it would fix the build).

### Issue link(s)

N/A

### Extra Info

There will not be a changelog entry, this does not affect customers or the SDK distribution.

**Comment by athomas** [comment]
LGTM, approving this myself to unblock the release process.

**Comment by athomas** [comment]
Note, these fixes are only needed to facilitate other fixes. Stable has pending cherry-picks so a release is needed there anyway. Beta doesn't need a release but I will land these fixes on beta as well to unblock future beta releases. This issue can be closed once the stable is out because it's not important to track this for the beta branch.

**Comment by athomas** [comment]
Merged (inconsequential for releases).
